### PR TITLE
feat: network request logger via URLProtocol with ring buffer

### LINKED ISSUE #39: feat: network request logger via URLProtocol with ring buffer
Part of #37

## Context

Captures HTTP network traffic metadata from all API providers (Zapper, Kraken, Binance, Coinbase, CoinGecko) without modifying any provider code. Uses a `URLProtocol` subclass registered on a custom `URLSessionConfiguration` that gets injected through existing `session:` init params. Stores entries in an in-memory ring buffer for later retrieval via the debug server's `/network/log` endpoint.

**Design spec:** `.dev/pm/specs/2026-04-11-debug-server.md`

## Complexity Hint

**Subsystems touched:** URLProtocol (Foundation), ring buffer data structure
**Independent stories:** 1
**Structural changes needed:** Yes — new `NetworkLogger.swift` with URLProtocol subclass and ring buffer
**Scope note:** None detected

## Task

Build a `URLProtocol` subclass (`NetworkLogger`) that intercepts HTTP requests and records metadata to a shared ring buffer:

1. Create `NetworkLogEntry` struct: timestamp, request URL, HTTP method, status code, response size in bytes, elapsed time, redacted headers (Authorization values replaced with `***`)
2. Create `NetworkLogBuffer` actor: fixed-size ring buffer (500 entries), append, query with optional `since` timestamp and `limit`
3. Create `NetworkLogger` URLProtocol subclass: in `startLoading()`, record request start time, forward to a real URLSession, on response record metadata, append to shared buffer, deliver data to client
4. Provide a factory method to create a debug-configured `URLSession`: `static func debugSession() -> URLSession` that returns a session with `NetworkLogger` registered in its `protocolClasses`

The ring buffer must be `Sendable` (actor-isolated) since URLProtocol callbacks run on arbitrary threads.

## Acceptance Criteria

- [ ] VERIFY: `NetworkLogger` URLProtocol intercepts requests made through `NetworkLogger.debugSession()`
- [ ] VERIFY: `NetworkLogEntry` captures URL, method, status code, response size, elapsed time, and timestamp
- [ ] VERIFY: Authorization header values are redacted to `***` in logged entries (e.g., `"Authorization": "Bearer ***"`)
- [ ] VERIFY: `NetworkLogBuffer` ring buffer holds max 500 entries and evicts oldest when full
- [ ] VERIFY: `NetworkLogBuffer.entries(since:limit:)` returns entries filtered by timestamp and capped by limit
- [ ] VERIFY: Request/response bodies are NOT captured (only metadata)
- [ ] VERIFY: `NetworkLogger.debugSession()` returns a `URLSession` that works identically to `.shared` for all network operations (providers produce the same results)
- [ ] VERIFY: Tests cover ring buffer overflow, timestamp filtering, header redaction, and entry creation

## Edge Cases

- Request that fails with network error (no response) — log entry with nil status code and error description
- Request with no Authorization header — no redaction needed, log headers as-is
- Multiple Authorization-style headers (`API-Key`, `API-Sign`) — redact all sensitive header values
- Ring buffer queried while being written to — actor isolation handles this
- Empty ring buffer queried — return empty array

### Files to Create
- `Sources/Portu/Debug/NetworkLogger.swift` — URLProtocol subclass, `NetworkLogEntry` struct, `NetworkLogBuffer` actor, `debugSession()` factory

### Approach
1. Write tests for `NetworkLogBuffer`: append, overflow eviction, `entries(since:limit:)` filtering
2. Write tests for header redaction logic
3. Implement `NetworkLogEntry` struct (Sendable, Codable for JSON serialization)
4. Implement `NetworkLogBuffer` actor with fixed-size array and head/tail indices
5. Implement `NetworkLogger` URLProtocol: override `canInit`, `canonicalRequest`, `startLoading`, `stopLoading`
6. In `startLoading`: record start time, create a private URLSession (without NetworkLogger to avoid recursion), forward request, on completion record entry and deliver to client
7. Implement `debugSession()` factory that creates `URLSessionConfiguration.default` with `protocolClasses = [NetworkLogger.self]`
8. Integration test: make a request through `debugSession()`, verify entry appears in buffer

### Patterns to Follow
- URLProtocol subclass pattern: use a private `URLSession` for the actual network call to avoid infinite recursion
- Actor for the ring buffer (thread-safe access from URLProtocol callbacks)
- Header redaction: check for known sensitive header names (`Authorization`, `API-Key`, `API-Sign`, `API-Secret`)

### Constraints
- Zero third-party dependencies
- Must compile under Swift 6.2 strict concurrency — `NetworkLogEntry` must be `Sendable`
- URLProtocol runs on arbitrary threads — all shared state via actor
- Must not capture request/response bodies (security: bodies may contain API keys as POST params)

### In Scope
- `NetworkLogger` URLProtocol subclass
- `NetworkLogEntry` struct with metadata fields
- `NetworkLogBuffer` actor (ring buffer, 500 entries)
- `debugSession()` factory method
- Header value redaction for sensitive headers

### Out of Scope
- Wiring the debug session into providers (sub-issue #5 — activation)
- The `/network/log` HTTP endpoint (sub-issue #3 or #4 will wire it up once the server skeleton exists)
- Response body capture
- Persistent storage of log entries

## Dependencies

- Blocked by: none (independent of server skeleton)
- Blocks: #37 sub-issue 5 (activation wires debug session into providers)

## Definition of Done

- [ ] Implementation complete
- [ ] Tests passing (unit + integration)
- [ ] PR approved and merged
- [ ] Epic checklist updated

### PR DESCRIPTION
**Pull request by rube-de**
## Summary

Closes #39 (part of #37)

- **`NetworkLogEntry`** — `Sendable`, `Codable` struct capturing URL, method, status code, response size, elapsed time, timestamp, and redacted headers
- **`NetworkLogBuffer`** — actor-isolated ring buffer (500 entries) with `entries(since:limit:)` query API
- **`NetworkLogger`** — `URLProtocol` subclass that intercepts requests via a private forwarding session, logs metadata, and delivers data transparently to callers
- **`debugSession()`** — factory method returning a drop-in `URLSession` with logging enabled
- Header redaction covers `Authorization`, `API-Key`, `API-Sign`, `API-Secret`, `Cookie`, `Set-Cookie` (case-insensitive)
- Response bodies are never captured — only byte count is recorded

## Test plan

- [x] `NetworkLogBufferTests` (10 tests) — append, overflow eviction, wrap-around, since/limit filtering, clear, edge cases
- [x] `NetworkLogEntryTests` (12 tests) — all sensitive headers redacted, case-insensitive matching, cookie headers, Codable round-trip, nil status code
- [x] `NetworkLoggerTests` (5 tests) — end-to-end URLProtocol interception, failed request logging, body-not-captured verification, multiple requests, identity check
- [x] Full test suite passes (186+ tests)
- [x] Build succeeds (Debug)
- [x] SwiftLint clean (zero errors)

## Summary by CodeRabbit

* **New Features**
  * Added network activity logging with per-request metadata (URL, method, status, response size, elapsed time, headers with sensitive values redacted) and in-app query/clear capabilities for debugging.

* **Tests**
  * Added comprehensive tests for logging, buffer behavior (capacity/ordering), header redaction, serialization, and end-to-end integration of the logger.

### CODE REVIEW

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+struct NetworkLogEntry: Codable, Identifiable {
+    let id: UUID
+    let timestamp: Date
+    let url: String
+    let method: String
+    let statusCode: Int?
+    let responseSizeBytes: Int
+    let elapsed: TimeInterval
+    let headers: [String: String]
+    var errorDescription: String? = nil
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:14]: Fixed: kept `var` with `= nil` default — `let` with default conflicts with `Codable` synthesis (Swift excludes it from the memberwise init). The struct is a value type so mutability is harmless.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 2]
```diff
@@ -0,0 +1,176 @@
+import Foundation
+
+// MARK: - NetworkLogEntry
+
+struct NetworkLogEntry: Codable, Identifiable {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:5]: Fixed: added explicit `Sendable` conformance. The formatter strips it (implicit in Swift 6 for value types), but the intent is documented.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    func clear() {
+        buffer = Array(repeating: nil, count: capacity)
+        writeIndex = 0
+        filled = 0
+    }
+}
+
+// MARK: - NetworkLogger
+
+final class NetworkLogger: URLProtocol, @unchecked Sendable {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:84]: Fixed: added safety rationale comment above the class declaration explaining URLProtocol's serial dispatch guarantee.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            url: request.url?.absoluteString ?? "unknown",
+            method: request.httpMethod ?? "GET",
+            statusCode: statusCode,
+            responseSizeBytes: responseData.count,
+            elapsed: elapsed,
+            headers: NetworkLogEntry.redactHeaders(headers),
+            errorDescription: error?.localizedDescription)
+
+        responseData = Data()
+        Task { await NetworkLogBuffer.shared.append(entry) }
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:178]: Fixed: replaced `Task.sleep(200ms)` in tests with deterministic polling via `waitForEntries(count:timeout:)`. The `Task { await ... }` dispatch is inherent to URLProtocol (delegate callbacks are synchronous) — the polling helper eliminates the timing dependency.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+// MARK: - NetworkLogBuffer
+
+actor NetworkLogBuffer {
+    static let shared = NetworkLogBuffer()
+
+    private var buffer: [NetworkLogEntry?]
+    private var writeIndex = 0
+    private var filled = 0
+    let capacity: Int
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift]: Fixed: replaced with `waitForEntries(count:timeout:)` that polls `entryCount` every 10ms with a 1s timeout.

[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift]: Fixed: changed to `private let capacity`.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    let id: UUID
+    let timestamp: Date
+    let url: String
+    let method: String
+    let statusCode: Int?
+    let responseSizeBytes: Int
+    let elapsed: TimeInterval
+    let headers: [String: String]
+    var errorDescription: String? = nil
+
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:15]: Fixed: explicit Sendable added (formatter strips it as implicit in Swift 6 value types).

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+// MARK: - NetworkLogger
+
+final class NetworkLogger: URLProtocol, @unchecked Sendable {
+    private static let handledKey = "NetworkLogger.handled"
+    private var internalSession: URLSession?
+    private var internalTask: URLSessionDataTask?
+    private var startTime: Date?
+    private var responseData = Data()
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift]: Fixed: replaced `responseData: Data` with `responseSizeCounter: Int`. No body data is ever held in memory now — only the byte count is incremented.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+    /// Test hook: protocol classes prepended to the internal forwarding session.
+    /// Globally registered protocols don't apply to URLSession(configuration:) sessions,
+    /// so tests inject a mock responder here instead.
+    nonisolated(unsafe) static var forwardingProtocolClasses: [AnyClass] = []
+
+    // swiftlint:disable:next static_over_final_class
+    override class func canInit(with request: URLRequest) -> Bool {
+        URLProtocol.property(forKey: handledKey, in: request) == nil
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift]: Fixed: `canInit` now guards on `scheme == "http" || scheme == "https"`.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+actor NetworkLogBuffer {
+    static let shared = NetworkLogBuffer()
+
+    private var buffer: [NetworkLogEntry?]
+    private var writeIndex = 0
+    private var filled = 0
+    let capacity: Int
+
+    init(capacity: Int = 500) {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:40]: Fixed: added `precondition(capacity > 0)` in the initializer.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+        var result: [NetworkLogEntry] = []
+        result.reserveCapacity(filled)
+        for i in 0 ..< filled {
+            let index = (start + i) % capacity
+            if let entry = buffer[index] {
+                if let since, entry.timestamp < since { continue }
+                result.append(entry)
+            }
+        }
+        if let limit { return Array(result.suffix(limit)) }
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift]: Fixed: added `guard let limit, limit > 0` before calling `suffix`.

File: Tests/PortuTests/NetworkLoggerTests.swift [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        let httpResponse = try #require(response as? HTTPURLResponse)
+
+        #expect(httpResponse.statusCode == 200)
+        #expect(data == LoggerTestProtocol.responseBody)
+
+        // Allow the async Task inside NetworkLogger to complete
+        try await Task.sleep(for: .milliseconds(200))
+
+        let entries = await NetworkLogBuffer.shared.entries()
+        #expect(entries.count == 1)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/NetworkLoggerTests.swift]: Fixed: replaced all `Task.sleep` calls with `waitForEntries(count:timeout:)` that polls `entryCount` every 10ms with 1s timeout.

File: Tests/PortuTests/NetworkLoggerTests.swift [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            _ = try await session.data(from: #require(URL(string: "https://test.local/fail")))
+            Issue.record("Expected network error")
+        } catch {
+            #expect(error is URLError)
+        }
+
+        try await Task.sleep(for: .milliseconds(200))
+
+        let entries = await NetworkLogBuffer.shared.entries()
+        #expect(entries.count == 1)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/NetworkLoggerTests.swift:105]: Fixed: replaced with deterministic polling.

File: Tests/PortuTests/NetworkLoggerTests.swift [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        defer { NetworkLogger.forwardingProtocolClasses = [] }
+        LoggerTestProtocol.responseBody = Data(repeating: 0xFF, count: 4096)
+
+        let session = NetworkLogger.debugSession()
+        _ = try await session.data(from: #require(URL(string: "https://test.local/large")))
+
+        try await Task.sleep(for: .milliseconds(200))
+
+        let entries = await NetworkLogBuffer.shared.entries()
+        #expect(entries.count == 1)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/NetworkLoggerTests.swift:120]: Fixed: replaced with deterministic polling.

File: Tests/PortuTests/NetworkLoggerTests.swift [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+        let session = NetworkLogger.debugSession()
+
+        for i in 0 ..< 3 {
+            _ = try await session.data(from: #require(URL(string: "https://test.local/req/\(i)")))
+        }
+
+        try await Task.sleep(for: .milliseconds(200))
+
+        let entries = await NetworkLogBuffer.shared.entries()
+        #expect(entries.count == 3)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/NetworkLoggerTests.swift:135]: Fixed: replaced with deterministic polling.

File: Sources/Portu/Debug/NetworkLogger.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift]: Fixed: replaced `responseData: Data` with an `Int` counter. No body data is ever held in memory.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+        guard let mutable = (request as NSURLRequest).mutableCopy() as? NSMutableURLRequest else {
+            client?.urlProtocol(self, didFailWithError: URLError(.badURL))
+            return
+        }
+        URLProtocol.setProperty(true, forKey: Self.handledKey, in: mutable)
+        let config = URLSessionConfiguration.default
+        if !Self.forwardingProtocolClasses.isEmpty {
+            config.protocolClasses = Self.forwardingProtocolClasses + (config.protocolClasses ?? [])
+        }
+        internalSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:124]: Answered: creating a per-instance URLSession is the standard URLProtocol forwarding pattern. The session is invalidated in both `stopLoading()` and `didCompleteWithError` — no leak. Apple's own documentation examples use this approach.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+        _: URLSession,
+        dataTask _: URLSessionDataTask,
+        didReceive response: URLResponse,
+        completionHandler: @escaping @Sendable (URLSession.ResponseDisposition) -> Void) {
+        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
+        completionHandler(.allow)
+    }
+
+    func urlSession(_: URLSession, dataTask _: URLSessionDataTask, didReceive data: Data) {
+        responseData.append(data)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift]: Fixed: now increments a byte counter (`responseSizeCounter += data.count`) instead of accumulating body data.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+
+// MARK: - URLSessionDataDelegate
+
+extension NetworkLogger: URLSessionDataDelegate {
+    func urlSession(
+        _: URLSession,
+        dataTask _: URLSessionDataTask,
+        didReceive response: URLResponse,
+        completionHandler: @escaping @Sendable (URLSession.ResponseDisposition) -> Void) {
+        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:151]: Answered: `.notAllowed` is intentional — the logger is a transparent passthrough and should not inject caching behavior. The original request's caching is handled by the system before our URLProtocol intercepts.

File: Tests/PortuTests/NetworkLoggerTests.swift [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+        request.setValue("Bearer test-token", forHTTPHeaderField: "Authorization")
+
+        let (data, response) = try await session.data(for: request)
+        let httpResponse = try #require(response as? HTTPURLResponse)
+
+        #expect(httpResponse.statusCode == 200)
+        #expect(data == LoggerTestProtocol.responseBody)
+
+        // Allow the async Task inside NetworkLogger to complete
+        try await Task.sleep(for: .milliseconds(200))
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/NetworkLoggerTests.swift]: Fixed: replaced with deterministic `waitForEntries(count:timeout:)` polling helper.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 19]
```diff
@@ -0,0 +1,181 @@
+import Foundation
+
+// MARK: - NetworkLogEntry
+
+struct NetworkLogEntry: Codable, Identifiable {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:5]: Fixed: explicit `Sendable` is added in source but SwiftFormat strips it (implicit for value types in Swift 6 strict concurrency). The conformance is enforced at compile time regardless.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+    let id: UUID
+    let timestamp: Date
+    let url: String
+    let method: String
+    let statusCode: Int?
+    let responseSizeBytes: Int
+    let elapsed: TimeInterval
+    let headers: [String: String]
+    var errorDescription: String? = nil
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:14]: Answered: `let errorDescription: String? = nil` causes Swift to exclude the property from the memberwise init when `Codable` is present (the compiler treats it as a constant with a fixed value). `var` with `= nil` is the pragmatic choice — the struct is a value type, so the field is effectively immutable after construction.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+            timestamp: startTime ?? Date(),
+            url: request.url?.absoluteString ?? "unknown",
+            method: request.httpMethod ?? "GET",
+            statusCode: statusCode,
+            responseSizeBytes: responseSizeCounter,
+            elapsed: elapsed,
+            headers: NetworkLogEntry.redactHeaders(headers),
+            errorDescription: error?.localizedDescription)
+
+        Task { await NetworkLogBuffer.shared.append(entry) }
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:178]: Answered: the `Task { await ... }` is unavoidable — `URLSessionDataDelegate` callbacks are synchronous, so there is no structured-concurrency seam to bridge. Tests now use deterministic polling (`waitForEntries`) instead of `Task.sleep`, eliminating the flakiness concern.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+        var result: [NetworkLogEntry] = []
+        result.reserveCapacity(filled)
+        for i in 0 ..< filled {
+            let index = (start + i) % capacity
+            if let entry = buffer[index] {
+                if let since, entry.timestamp < since { continue }
+                result.append(entry)
+            }
+        }
+        guard let limit, limit > 0 else { return result }
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift]: Fixed: split the guard — `limit: 0` now returns `[]` instead of falling through to return all entries.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+// MARK: - URLSessionDataDelegate
+
+extension NetworkLogger: URLSessionDataDelegate {
+    func urlSession(
+        _: URLSession,
+        dataTask _: URLSessionDataTask,
+        didReceive response: URLResponse,
+        completionHandler: @escaping @Sendable (URLSession.ResponseDisposition) -> Void) {
+        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
+        completionHandler(.allow)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:152]: Answered: `.notAllowed` is intentional — the logger is a transparent passthrough that must not inject caching behavior. The original request's cache policy is honored by the internal forwarding session.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+    }
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:153]: Answered: redirects are handled by the internal `URLSession` (line 113). The delegate receives the final response after all redirects. Since we don't implement `urlSession(_:task:willPerformHTTPRedirection:...)`, the default behavior (follow redirects) applies — same as `URLSession.shared`.

File: Tests/PortuTests/NetworkLoggerTests.swift [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+
+private func waitForEntries(
+    in buffer: NetworkLogBuffer = .shared,
+    count: Int,
+    timeout: TimeInterval = 1.0) async throws {
+    let deadline = Date.now.addingTimeInterval(timeout)
+    while await buffer.entryCount < count, Date.now < deadline {
+        try await Task.sleep(for: .milliseconds(10))
+    }
+}
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/NetworkLoggerTests.swift:53]: Answered: the polling helper is best-effort — if entries don't appear in time, the immediately following `#expect(entries.count == N)` assertion fails with a clear message. Adding a throw would duplicate the failure signal without improving diagnostics.

[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/NetworkLoggerTests.swift:53]: Answered: same rationale as above — the `#expect` assertions after the wait provide the actual failure signal. A throw on timeout would add noise without improving test diagnostics since the root cause (missing entry) is surfaced by the count assertion.

**Comment by rube-de** [comment]
> **Claude finished @rube-de's task in 3m 29s** — Code Review — PR #44: Network Request Logger...

Thanks for the thorough review. Addressed in dcb6e5f:

**1. `waitForEntries` timeout (Must Fix)** — Fixed. Added `Issue.record` after the polling loop with expected/actual counts and timeout duration for CI diagnostics.

**2. `var errorDescription` (Must Fix)** — Respectfully disagree. The SE-0242 citation is incorrect for `let` properties: SE-0242 specifies that `let` properties with default values are *excluded* from the memberwise initializer (only `var` properties get the default-parameter treatment). Verified empirically — `let errorDescription: String? = nil` causes "extra argument 'errorDescription' in call" because the compiler omits the parameter entirely. The alternative `let errorDescription: String?` (no default) requires updating 15+ call sites to pass `errorDescription: nil` explicitly. `var` on a value-type struct is harmless — the field is effectively immutable after construction.

**3. SwiftFormat `redundantConformances` (Should Fix)** — Good root-cause diagnosis. Modifying `.swiftformat` rules is a project-wide decision outside the scope of this feature PR — will address separately.

**4. `#if DEBUG` on `forwardingProtocolClasses` (Should Fix)** — Fixed. Both the property and its usage in `startLoading` are now wrapped in `#if DEBUG`.

**5. Task ordering comment (Nit)** — Fixed. Added two-line comment explaining the sync→async bridge and pointing to `waitForEntries()`.

**6. `debugSession()` comment (Nit)** — Fixed. Added comment clarifying the intentional replace-not-prepend.

**7. Codable round-trip test (Nit)** — Fixed. Added assertions for `timestamp`, `elapsed`, `headers`, and `errorDescription`.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 26]
```diff
@@ -0,0 +1,182 @@
+import Foundation
+
+// MARK: - NetworkLogEntry
+
+struct NetworkLogEntry: Codable, Identifiable {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:5]: Answered: `Sendable` conformance is implicit in Swift 6 for value types with all-Sendable stored properties — the compiler enforces it at every actor boundary. SwiftFormat's `redundantConformances` rule strips the explicit declaration. A project-wide `.swiftformat` config change is needed to preserve it; that's outside scope of this PR.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+    let id: UUID
+    let timestamp: Date
+    let url: String
+    let method: String
+    let statusCode: Int?
+    let responseSizeBytes: Int
+    let elapsed: TimeInterval
+    let headers: [String: String]
+    var errorDescription: String? = nil
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:14]: Answered: `NetworkLogEntry` is implicitly `Sendable` in Swift 6 — all stored properties are `Sendable` value types. The compiler enforces this at every actor boundary. SwiftFormat's `redundantConformances` rule strips the explicit declaration; a project-wide config change is tracked separately.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+        URLProtocol.setProperty(true, forKey: Self.handledKey, in: mutable)
+        let config = URLSessionConfiguration.default
+        #if DEBUG
+            if !Self.forwardingProtocolClasses.isEmpty {
+                config.protocolClasses = Self.forwardingProtocolClasses + (config.protocolClasses ?? [])
+            }
+        #endif
+        internalSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)
+        internalTask = internalSession?.dataTask(with: mutable as URLRequest)
+        internalTask?.resume()
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:126]: Answered: the rationale comment above the class declaration (lines 83–84) explains this: URLProtocol serializes `startLoading`, `stopLoading`, and all `URLSessionDataDelegate` callbacks on its own private serial queue. The mutable instance state is single-threaded by the framework's guarantee — `@unchecked Sendable` is the standard pattern for URLProtocol subclasses.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 29]
```diff
@@ -0,0 +1,189 @@
+import Foundation
+
+// MARK: - NetworkLogEntry
+
+struct NetworkLogEntry: Codable, Identifiable {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:5]: Answered: agreed — tracked as a project-wide .swiftformat config change (outside this PR's scope). The compiler enforces Sendable at every actor boundary regardless of the explicit declaration.

File: Sources/Portu/Debug/NetworkLogger.swift (hunk 27 above)
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:14]: Answered: tested empirically — `let errorDescription: String? = nil` on a `Codable` struct causes the compiler to emit 'immutable property will not be decoded because it is declared with an initial value which cannot be overwritten' and then excludes the parameter from the memberwise init. `let errorDescription: String?` (no default) works but requires updating every call site. The `var` with `= nil` default is the pragmatic choice for a value-type struct.

File: Sources/Portu/Debug/NetworkLogger.swift [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+        self.buffer = Array(repeating: nil, count: capacity)
+    }
+
+    func append(_ entry: NetworkLogEntry) {
+        buffer[writeIndex] = entry
+        writeIndex = (writeIndex + 1) % capacity
+        filled = min(filled + 1, capacity)
+    }
+
+    func entries(since: Date? = nil, limit: Int? = nil) -> [NetworkLogEntry] {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Debug/NetworkLogger.swift:52]: Answered: the filter `entry.timestamp < since` means entries with `timestamp >= since` are included — i.e., 'give me everything from this point forward' which is the standard 'since' semantic (same as HTTP `If-Modified-Since`). The name + behavior are self-consistent.
